File: problem-08/c/main.c
#include <stdio.h>
#include <stdlib.h>

long long int array_product(int n, int *array) {
	long long int prod = 1;
	for (int i = 0; i < n; i++) {
		prod *= array[i];
	}
	return prod;
}

long long int max_int(long long int a, long long int b) {
	if (a < b) return b;
	return a;
}

long long int solution(int n, const char *path) {
	// Open File
	FILE *fp = NULL;
	if ((fp = fopen(path, "r")) == NULL) {
		printf("fatal - failed to open file (%s)\n", path);
		return 0;
	}

	// Search
	int *numbers = malloc(n * sizeof(int));
	int idx = 0;
	int input = 0;
	long long int output = 0;
	while ((fscanf(fp, "%1d", &input)) != EOF) {
		// Load Array
		numbers[idx % n] = input;

		// Check Sum
		if (!(idx < n)) {
			output = max_int(output, array_product(n, numbers));
		}

		idx++;
	}

	// Close File
	fclose(fp);

	return output;
}

int main() {
	const char *path = "/home/knight/Documents/github/project-euler/problem-08/input.txt";
	printf("Solution #8: %lli\n", solution(13, path));
}

File: problem-08/rust/src/main.rs
use std::fs;

fn main() {
    let path: &str = "/home/knight/Documents/github/project-euler/problem-08/input.txt";
    println!("Solution #8: {}", solution(13, path));
}

fn solution(n: i32, path: &str) -> u64 {
    return fs::read_to_string(path)
        .unwrap()
        .chars()
        .filter_map(|c| c.to_digit(10))
        .map(|x| x as u64)
        .collect::<Vec<_>>()
        .windows(n as usize)
        .map(|window| window.iter().product())
        .max()
        .unwrap()
}
